src/app/components/import/ImportingModalContent.js

```diff
@@ -1,115 +1,221 @@
-        }
-        setStep(ENCRYPTING);
-    };
-
-    const encryptContacts = async () => {
-        const publicKey = privateKey.toPublic();
-        for (const vcardContact of vcardContacts) {
-            try {
-                const contactEncrypted = await prepareContact(vcardContact, privateKey, publicKey);
-                setTrack((track) => ({ ...track, encrypted: [...track.encrypted, contactEncrypted] }));
-            } catch (error) {
-                setTrack((track) => ({ ...track, failedOnEncrypt: [...track.failed, vcardContact] }));
+
+            return encryptedContacts;
+        };
+
+        /*
+			Send a batch of contacts to the API
+		*/
+        const submitBatch = async ({ contacts = [], labels }, { signal }) => {
+            if (signal.aborted || !contacts.length) {
+                return;
             }
-        }
-        onEncryptingDone();
-        setStep(IMPORTING);
-    };
-
-    const importContacts = async (contacts) => {
-        // split encrypted contacts depending on having the CATEGORIES property
-        const withCategories = contacts.filter(({ Cards }) =>
-            Cards.some(({ Type, Data }) => Type === CLEAR_TEXT && Data.includes('CATEGORIES'))
-        );
-        const withoutCategories = contacts.filter(({ Cards }) =>
-            Cards.every(({ Type, Data }) => Type !== CLEAR_TEXT || !Data.includes('CATEGORIES'))
-        );
-
-        // send encrypted contacts to API
-        const responses = (await api(
-            addContacts({ Contacts: withCategories, Overwrite: OVERWRITE_CONTACT, Labels: INCLUDE })
-        )).Responses.concat(
-            (await api(addContacts({ Contacts: withoutCategories, Overwrite: OVERWRITE_CONTACT, Labels: IGNORE })))
-                .Responses
-        ).map(({ Response }) => Response);
-
-        const imported = responses.reduce((acc, { Code }) => {
-            if (Code === SUCCESS_IMPORT_CODE) {
-                return acc + 1;
+
+            const indexMap = contacts.map(({ index }) => index);
+
+            const responses = (await apiWithAbort(
+                addContacts({
+                    Contacts: contacts.map(({ contact }) => contact),
+                    Overwrite: OVERWRITE_CONTACT,
+                    Labels: labels
+                })
+            )).Responses.map(({ Response }) => Response);
+
+            const { submittedBatch, failedOnSubmitBatch } = responses.reduce(
+                (acc, { Code, Error, Contact: { ID } }, i) => {
+                    const index = indexMap[i];
+                    if (Code === SUCCESS_IMPORT_CODE) {
+                        acc.submittedBatch.push(ID);
+                    } else {
+                        acc.failedOnSubmitBatch.push({ index, message: createSubmitErrorMessage(index, Error) });
+                    }
+                    return acc;
+                },
+                { submittedBatch: [], failedOnSubmitBatch: [] }
+            );
+            !signal.aborted &&
```

Comment:
perhaps more clear with `if (signal.aborted) { return; }` above `const { submittedBatch, failedOnSubmitBatch } = responses.reduce(`?

Reply:
I wonder if it's possible (although clearly highly unlikely) that `signal.aborted` becomes `true` during the execution of the `reduce`. In such a case, `setModel` could be called after the component is unmounted, triggering a React complain

Reply:
Shouldn't be possible since it's a sync op